components/ShadowLayer/src/MDCShadowLayer.m

```diff
@@ -304,25 +334,77 @@ - (void)commonLayoutSublayers {
   _topShadow.bounds = bounds;
 
   if (_shadowMaskEnabled) {
-    _bottomShadow.mask = [self shadowLayerMaskForLayer:_bottomShadow];
-    _topShadow.mask = [self shadowLayerMaskForLayer:_topShadow];
+    [self configureShadowLayerMaskForLayer:_topShadowMask];
+    [self configureShadowLayerMaskForLayer:_bottomShadowMask];
   }
   // Enforce shadowPaths because otherwise no shadows can be drawn. If a shadowPath
   // is already set, use that, otherwise fallback to just a regular rect because path.
-  if (!_bottomShadow.shadowPath) {
+  if (!_bottomShadow.shadowPath || _shadowPathIsInvalid) {
     if (self.shadowPath) {
       _bottomShadow.shadowPath = self.shadowPath;
     } else {
       _bottomShadow.shadowPath = [self defaultShadowPath].CGPath;
     }
   }
-  if (!_topShadow.shadowPath) {
+  if (!_topShadow.shadowPath || _shadowPathIsInvalid) {
     if (self.shadowPath) {
       _topShadow.shadowPath = self.shadowPath;
     } else {
       _topShadow.shadowPath = [self defaultShadowPath].CGPath;
     }
   }
+  _shadowPathIsInvalid = NO;
+}
+
+@end
+
+@implementation MDCAnimatedShapeLayer
+
+- (id<CAAction>)actionForKey:(NSString *)key {
+  if ([key isEqualToString:@"path"] || [key isEqualToString:@"shadowPath"]) {
+    CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:key];
+    animation.fromValue = [self.presentationLayer valueForKey:key];
+
+    // We have to create a pending animation because if we are inside a UIKit animation block we
+    // won't know any properties of the animation block until it is commited.
+    MDCPendingAnimation *pendingAnim = [[MDCPendingAnimation alloc] init];
+    pendingAnim.animation = animation;
+
+    return pendingAnim;
+  }
+  return [super actionForKey:key];
+}
+
+@end
+
+@implementation MDCPendingAnimation
+
+- (void)runActionForKey:(NSString *)event object:(id)anObject arguments:(NSDictionary *)dict {
+  if ([anObject isKindOfClass:[MDCAnimatedShapeLayer class]]) {
+    MDCAnimatedShapeLayer *layer = (MDCAnimatedShapeLayer *)anObject;
+
+    // In order to synchronize our animation with UIKit animations we have to fetch the resizing
+    // animation created by UIKit and copy the configuration to our custom animation.
+    CALayer *animatedLayer = layer.animationParentLayer;
+    CAAnimation *boundsAction = [animatedLayer animationForKey:@"bounds.size"];
+    if ([boundsAction isKindOfClass:[CABasicAnimation class]]) {
+      [self configureAnimation:(CABasicAnimation *)boundsAction];
+      [layer addAnimation:self.animation forKey:event];
+    }
+  }
+}
+
+- (void)configureAnimation:(CAAnimation *)template {
+  self.animation.autoreverses = template.autoreverses;
+  self.animation.beginTime = template.beginTime;
```

Comment:
This line may prove to be problematic if you intend to support delayed animations. Unfortunately, UIKit implements delayed animations using a hidden API on CAAnimation ([beginTimeMode](https://github.com/JaviSoto/iOS10-Runtime-Headers/blob/master/Frameworks/QuartzCore.framework/CAAnimation.h#L15)), which they're able to set to "relative". When you set it, however, it'll get set to "absolute", which will almost always be incorrect.

So here you have to either offset the time to match the layer's time, or copy the animation so that you can inherit the private details of beginTime.

```objc
begimTime = [layer convertTime:CACurrentMediaTime() fromLayer:nil] + beginTime
```

Reply:
I've changed the implementation to copy the animation instance.